find auto-generated secret_key_base in development

gencer opened the pull request:
With this fix, we will try latest changes in Rails 5.2 together with standard auto-generated secret_key_base in development as a fallback.

If no specified key found, auto-generated value will be used instead.

This fixes #4864

gencer commented:
@tegon failed job is something about networking and/or i/o side on Travis. Maybe reschedule will fix the problem.

> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
> The build has been terminated

tegon commented:
@gencer thanks! Can you add a new test case for this?

gencer commented:
@tegon I've added test case for `secret_key_base`, squashed them and pushed back.

ttilberg commented:
@tegon @gencer I referenced that branch from my Gemfile and it worked as expected.

The only thing I would suggest is the following consideration:

It seems to be that the Rails team views `Rails.application.secret_key_base` as the authoritative source for this information based on the fact that this method [already aggregates](https://github.com/rails/rails/blob/master/railties/lib/rails/application.rb#L430) the [other configuration locations](https://github.com/rails/rails/blob/master/railties/lib/rails/application.rb#L401), in addition to the new dev/test default. As such, I think this should be the first location checked, with the others being the fallbacks to preserve backwards compatibility. 

Even in the case where it does have to fall back, it's fewer checks and iterations happening as it's not going to be responding to that method in the first place.

Either way, this is successful in Rails 5.2.

Thanks for doing the PR on this. I understood what needed to be done, but am a little intimidated of the testing side of it.

tegon commented on test/secret_key_finder_test.rb:
```diff
@@ -32,6 +32,25 @@ def config
+  end
+  
+  def secret_key_base
+    'secret_key_base' if Rails.env.development? || Rails.env.test?
```
Why do we need this `if` statement here?

gencer replied:
Oh! How did I missed that! I was testing something else and I just left them there. I'll fix and squash now.

gencer commented:
@tegon the line you were referring was absolutely unnecessary as its already in dev/test environment. I removed that in addition, I had one more blank line which I removed that too.

Squashed&Pushed now.

issei-m commented:
Where is this fix released? I'm currently facing on this same problem on Devise 4.4.3.
Thanks.

tegon commented:
@issei-m it was released in `4.4.3`. Please move the discussion to #4879, which is an open issue where we're checking this. Also, it'd be good if you provide us your Rails version and the output of the following commands:
```
Rails.application.credentials.secret_key_base

Rails.application.secrets.secret_key_base

Rails.application.config.secret_key_base

issei-m commented:
@tagon Oh really? It seems to me, `v4.4.3` [has no commits introduced in this PR](https://github.com/plataformatec/devise/blob/v4.4.3/lib/devise/secret_key_finder.rb#L10-L16).
And I cloned this repository and I hit the some commands to know who has these commits right now, then only `master` branch seems to have, while v4.4.3 doesn't:

```
$ git branch --contains 6c91648
* master

$ git tag --contains 6c91648

```

tegon commented:
@issei-m Ouch, you're right. I've mistaken this with https://github.com/plataformatec/devise/commit/962cea2039c72a92691af734ebbd8495dd5c0501, my bad. Thanks for pointing that out.
I'll see which commits we have unreleased and if there's something else to include in the next release and then bump a new version.
Just to make sure, if you use Devise from master it works, right?

tegon commented:
@issei-m thanks, I'll prepare a release as soon as I can.

tegon commented:
Released in `v4.5.0`

swrobel commented:
@gencer @tegon what about [ttilberg's comment](https://github.com/plataformatec/devise/pull/4869#issuecomment-387874576) which points out the incongruity between Devise behavior & Rails.

For example, on Heroku, they set `ENV['SECRET_KEY_BASE']` if it isn't already set for Rails apps. As a result, Rails ends up using that key instead of `Rails.application.credentials.secret_key_base`, although Devise will prefer the latter, resulting in different values for each.

issei-m commented:
@swrobel Simply, you can remove `secret_key_base` from your credentials.yml.inc, so `ENV['SECRET_KEY_BASE']` will be used always in Heroku environment.

swrobel commented:
@issei-m I still don't get why devise should behave differently from Rails

tegon commented:
@swrobel Devise's `secret_key_base` doesn't necessarily need to be the same as Rails's. We only try to get it from Rails for convenience, so that you don't need to manually set it to get started.
When I looked at @ttilberg comment, I thought it was only a matter of order - e.g. where we look first for a key - I didn't realize that Rails looked for the environment variable too.
We can support this in a new version, but this PR was mostly about fixing a bug we introduced in a previous version. Would you mind opening an issue asking to support this?

ttilberg commented:
@tegon My comment wasn't as much about "where to look first" for the sake of looking in many places, but instead a solution to leverage the configuration of modern rails apps while maintaining backwards compatibility. There's a few links to the relevant parts of the rails source in both my comment above and the original issue this closes that I thought were helpful to understand the "current rails configuration strategy" -- it felt a little all over the place but seems to get unified at `Rails.application.secret_key_base` accounting for all expected possibilities, including the new dynamic generation for dev/test.

What is the best way to ask for this support in a new issue? Something along the lines of "Feature: Prefer Rails default SECRET_KEY_BASE configuration?" It feels a bit superfluous now, but I think it's a good future-proofing thing to do.

gencer commented:
What I get so far that we need to also check `ENV` for `SECRET_KEY_BASE` (if possible at first) and regardless what we've found so far, we will use this Environment variable as default secret key.

Simply, `ENV["SECRET_KEY_BASE"]` has higher priority and overrides all (or If found, no need to check further places)

We need an another **PR** for this. I'll have a look in spare time (No promises) and create PR with tests (if no one contribute in this time of course). But, We still need an issue to track those changes.

ttilberg commented:
Hey @gencer -- I was already in the middle of taking a stab at refactoring the key finder just now, I'll add that consideration (to check ENV first) as well and post a PR in a moment here. In addition, it was about checking the correct "default location in Rails" first, instead of last. Thanks for your help working through this!

ttilberg commented:
@tegon @gencer Can you guys take a look at PR I just posted above? I don't have a lot of experience/confidence with contributing to such a serious project, so I'd appreciate feedback on:

- Was this an appropriate way to contribute (going straight to PR without an issue, etc)
- The refactor decisions

I added a test for the `ENV` addition @gencer mentioned, and all tests are still passing.

md5 commented:
> @swrobel Simply, you can remove `secret_key_base` from your credentials.yml.inc, so `ENV['SECRET_KEY_BASE']` will be used always in Heroku environment.

This is not the case. If `Rails.application.credentials` exists at all, Devise will try to get `secret_key_base` from `Rails.application.credentials.secret_key_base`. If it does not exist, it doesn't try the other sources. So if you want to set `SECRET_KEY_BASE` from an environment variable and still use `config/credentials.yml.enc` for other values, that won't work.

I think @ttilberg is correct that `Rails.application.secret_key_base` should be called first if it exists to avoid this problem.

md5 commented:
Or rather, removing `secret_key_base` from `credentials.yml.enc` causes `ENV['SECRET_KEY_BASE']` to be used by Rails, but not by Devise.
